# Data Science Jobs Scraper

In [74]:
import re
import requests
import numpy as np
import pandas as pd
from bs4 import BeautifulSoup

### Import Target Keywords

In [98]:
df = pd.read_csv("ds_skills.csv")

df["Skill"] = df["Skill"].str.lower()
df["Category"] = df["Category"].str.lower()
df['total_jobs'] = 0

print("Total Skills: ", df.shape[0], " technologies")
df.head()

Total Skills:  50  technologies


Skill     Category  total_jobs
0      python  programming           0
1           r  programming           0
2         sql     database           0
3        java  programming           0
4  javascript  programming           0

In [76]:
skills = df['Skill'].to_list()

print(skills)

['python', 'r', 'sql', 'java', 'javascript', 'c', 'scala', 'julia', 'sas', 'matlab', 'matplotlib', 'seaborn', 'bokeh', 'plotly', 'ggplot', 'numpy', 'scipy', 'pandas', 'scikit learn', 'keras', 'pytorch', 'xgboost', 'tensorflow', 'aws', 'azure', 'haskel', 'swift', 'octave', 'perl', 'lisp', 'excel', 'power bi', 'tableau', 'google cloud', 'spark', 'hadoop', 'git', 'spss', 'stata', 'pig', 'hbase', 'docker', 'kubernetes', 'statsmodels', 'nltk', 'scrapy', 'beautiful soup', 'selenium', 'github', 'nosql']


## Indeed

In [88]:
url = f'https://www.indeed.com/jobs?q=%22data+scientist%22&l=United+States'

try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('div', id="searchCountPages").get_text()
    numb = count_str.split()
    indeed_total = int(re.sub(",", "", numb[-2]))

except Exception as e:
    print(f'error: {e}')

print(indeed_total, "Data Science Jobs")

4773 Data Science Jobs


Skill     Category  total_jobs
0      python  programming        4773
1           r  programming        4773
2         sql     database        4773
3        java  programming        4773
4  javascript  programming        4773

In [77]:
indeed_list = []

for skill in skills:
    url = f'https://www.indeed.com/jobs?q=%22data+scientist%22+%22{skill}%22&l=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('div', id="searchCountPages").get_text()
        numb = count_str.split()
        indeed_list.append(numb[-2])
    except Exception as e:
        print(f'error: {e}')
        
print(indeed_list)

['3,640', '2,474', '2,629', '1,070', '276', '425', '645', '65', '723', '449', '128', '40', '84', '56', '87', '307', '156', '363', '426', '273', '406', '65', '683', '1,107', '487', '1', '10', '40', '106', '2', '507', '241', '1,081', '179', '1,238', '885', '259', '165', '62', '143', '105', '281', '188', '35', '42', '2', '1', '7', '172', '323']


In [99]:
indeed = []

for i in indeed_list:
    indeed.append(int(re.sub(",", "", i)))

df['total_jobs'] = df['total_jobs'] + indeed_total
df['indeed'] = indeed
df.head()

Skill     Category  total_jobs  indeed
0      python  programming        4773    3640
1           r  programming        4773    2474
2         sql     database        4773    2629
3        java  programming        4773    1070
4  javascript  programming        4773     276

## Simply Hired

In [93]:
url = f'https://www.simplyhired.com/search?q=%22data+scientist%22&l=United+States'
    
try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('span', class_="CategoryPath-total").get_text()
    simply_total = int(re.sub(",", "", count_str))

except Exception as e:
    print(f'error: {e}')

print(simply_total, "Data Science Jobs")

3518 Data Science Jobs


In [96]:
simply_list = []

for skill in skills:
    url = f'https://www.simplyhired.com/search?q=%22data+scientist%22+%22{skill}%22&l=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('span', class_="CategoryPath-total").get_text()
        simply_list.append(count_str)
    except Exception as e:
        print(f'error: {e}')
        
print(simply_list)

['2,707', '1,814', '1,986', '765', '179', '325', '453', '60', '534', '319', '94', '29', '46', '53', '72', '227', '122', '293', '288', '214', '303', '47', '526', '788', '392', '1', '9', '13', '91', '2', '411', '201', '755', '121', '926', '690', '218', '131', '55', '108', '87', '191', '155', '24', '38', '2', '1', '7', '137', '265']


In [100]:
simply = []

for i in simply_list:
    simply.append(int(re.sub(",", "", i)))

df['total_jobs'] = df['total_jobs'] + simply_total
df['simply'] = simply
df.head()

Skill     Category  total_jobs  indeed  simply
0      python  programming        8291    3640    2707
1           r  programming        8291    2474    1814
2         sql     database        8291    2629    1986
3        java  programming        8291    1070     765
4  javascript  programming        8291     276     179

## Glassdoor

In [110]:
url = f'https://www.glassdoor.com/Job/jobs.htm?sc.keyword="data+scientist"+"%22python%22"'

r = requests.get(url)
print(r.status_code)

403


**Status code is 403 (Forbidden)**. Therefore, we can not scrape content from Glassdoor. Optionally, we can try using **Selenium** in the future.

## Zip Recruiter

In [115]:
url = f'https://www.ziprecruiter.com/candidate/search?search=data+scientist&location=United+States'

r = requests.get(url)
print(r.status_code)

403


**Status code is 403 (Forbidden)**. Therefore, we can not scrape content from Glassdoor. Optionally, we can try using **Selenium** in the future.

## Career Builder

In [124]:
url = 'https://www.careerbuilder.com/jobs?keywords=Data+Scientist&location=United+States'

try:
    r = requests.get(url)
    soup = BeautifulSoup(r.text, 'html.parser')
    count_str = soup.find('h1', class_="fz1rem").get_text()
    career_total = int(re.sub(",", "", count_str.split()[0]))

except Exception as e:
    print(f'error: {e}')

print(career_total, "Data Science Jobs")

3430 Data Science Jobs


In [126]:
career_list = []

for skill in skills:
    keyword = str(re.sub(",", "+", skill))
    url = f'https://www.careerbuilder.com/jobs?keywords={keyword}+Data+Scientist&location=United+States'
    
    try:
        r = requests.get(url)
        soup = BeautifulSoup(r.text, 'html.parser')
        count_str = soup.find('h1', class_="fz1rem").get_text()
        career_list.append(int(re.sub(",", "", count_str.split()[0])))
    
    except Exception as e:
        career_list.append(0)
        print(f'error: {e}')
        print(skill)
        
print(career_list)

error: 'NoneType' object has no attribute 'get_text'
bokeh
error: 'NoneType' object has no attribute 'get_text'
haskel
error: 'NoneType' object has no attribute 'get_text'
scrapy
error: 'NoneType' object has no attribute 'get_text'
beautiful soup
[901, 1091, 696, 363, 106, 495, 173, 4, 174, 107, 30, 12, 0, 6, 25, 71, 31, 68, 54, 65, 48, 15, 114, 285, 169, 0, 1, 1, 53, 1, 575, 52, 291, 46, 313, 278, 109, 23, 11, 63, 37, 117, 74, 3, 14, 0, 0, 8, 47, 106]


In [129]:
career = []

for i in career_list:
    career.append(int(i))

df['total_jobs'] = df['total_jobs'] + career_total
df['career'] = career
df.head()

Skill     Category  total_jobs  indeed  simply  career
0      python  programming       11721    3640    2707     901
1           r  programming       11721    2474    1814    1091
2         sql     database       11721    2629    1986     696
3        java  programming       11721    1070     765     363
4  javascript  programming       11721     276     179     106

### Export Dataset

In [130]:
print(df.shape)
df.to_csv("ds_scraped_data.csv")

(50, 6)
